core/state-machine/src/backend.rs

```diff
@@ -326,96 +379,99 @@ impl<H: Hasher> Backend<H> for InMemory<H> {
-		self.inner.get(&None)
-			.into_iter()
-			.flat_map(|map| map.iter().map(|(k, v)| (k.clone(), v.clone())))
-			.collect()
+		self.inner.0.iter().map(|(k, v)| (k.clone(), v.clone())).collect()
 	}
 
 	fn keys(&self, prefix: &[u8]) -> Vec<Vec<u8>> {
-		self.inner.get(&None)
-			.into_iter()
-			.flat_map(|map| map.keys().filter(|k| k.starts_with(prefix)).cloned())
-			.collect()
-	}
-
-	fn child_keys(&self, storage_key: &[u8], prefix: &[u8]) -> Vec<Vec<u8>> {
-		self.inner.get(&Some(storage_key.to_vec()))
-			.into_iter()
-			.flat_map(|map| map.keys().filter(|k| k.starts_with(prefix)).cloned())
-			.collect()
+		self.inner.0.keys().filter(|k| k.starts_with(prefix)).cloned().collect()
 	}
 
 	fn as_trie_backend(&mut self)-> Option<&TrieBackend<Self::TrieBackendStorage, H>> {
+		// Warning tihs do not use keyspace for child trie, that is not an issue unless
+		// the memorydb gets incorectly use to fead keyvalue database.
 		let mut mdb = MemoryDB::default();
-		let mut root = None;
 		let mut new_child_roots = Vec::new();
-		let mut root_map = None;
-		for (storage_key, map) in &self.inner {
-			if let Some(storage_key) = storage_key.as_ref() {
-				let ch = insert_into_memory_db::<H, _>(&mut mdb, map.clone().into_iter())?;
-				new_child_roots.push((storage_key.clone(), ch.as_ref().into()));
-			} else {
-				root_map = Some(map);
-			}
-		}
-		// root handling
-		if let Some(map) = root_map.take() {
-			root = Some(insert_into_memory_db::<H, _>(
+		let (root_map, child_map) = std::mem::replace(&mut self.inner, Default::default());
+
+		for (_k, (map, child_trie)) in child_map.into_iter() {
+			let ch = insert_into_memory_db::<H, _>(
 				&mut mdb,
-				map.clone().into_iter().chain(new_child_roots.into_iter())
-			)?);
+				map.into_iter(),
+				Some(child_trie.node_ref()),
+			)?;
+			new_child_roots.push((
+				child_trie.parent_trie().to_vec(),
+				child_trie.encoded_with_root(ch.as_ref())));
 		}
-		let root = match root {
-			Some(root) => root,
-			None => insert_into_memory_db::<H, _>(&mut mdb, ::std::iter::empty())?,
-		};
+		// root handling
+		let root = insert_into_memory_db::<H, _>(
+			&mut mdb,
+			root_map.into_iter().chain(new_child_roots.into_iter()),
+			None,
+		)?;
 		self.trie = Some(TrieBackend::new(mdb, root));
 		self.trie.as_ref()
 	}
 }
 
 /// Insert input pairs into memory db.
-pub(crate) fn insert_into_memory_db<H, I>(mdb: &mut MemoryDB<H>, input: I) -> Option<H::Out>
+pub(crate) fn insert_into_memory_db<H, I>(
+	mdb: &mut MemoryDB<H>,
+	input: I,
+	child_trie: Option<ChildTrieReadRef>
+) -> Option<H::Out>
 	where
 		H: Hasher,
 		I: IntoIterator<Item=(Vec<u8>, Vec<u8>)>,
 {
 	let mut root = <H as Hasher>::Out::default();
 	{
-		let mut trie = TrieDBMut::<H>::new(mdb, &mut root);
-		for (key, value) in input {
-			if let Err(e) = trie.insert(&key, &value) {
-				warn!(target: "trie", "Failed to write to trie: {}", e);
-				return None;
+		if let Some(child_trie) = child_trie.as_ref() {
```

Comment:
Reduce duplication:

```rust
let keyspace = child_trie.map(|child| child.keyspace());
let mut mdb = KeySpacedDBMut::new(&mut *mdb, keyspace);
```

Reply:
~~Note that it makes it harder to audit that keyspacedb is use for every triedb (some changes in trie crates could allow removing keyspacedb in the future: there is an allocation to skip here), but I like it better anyway :)~~
Aborting change due to lifetime issue.

Reply:
Sorry, I should have checked this first. This does work with the `as_ref`:

```rust
let keyspace = child_trie.as_ref().map(|child| child.keyspace());
let mut mdb = KeySpacedDBMut::new(&mut *mdb, keyspace);
```

Reply:
Er, you can get rid if the `if` statement too... the `if` and `else` are identical.

Reply:
The first comment suddenly make a lot more sense :+1: